File: Vanta Speech/Vanta Speech/Core/Calendar/OutlookCalendarManager.swift
import Combine
import Foundation
import UIKit

/// Главный координатор интеграции с Outlook Calendar
@MainActor
final class OutlookCalendarManager: ObservableObject {

    // MARK: - Singleton

    static let shared = OutlookCalendarManager()

    // MARK: - Dependencies

    let authManager: MSALAuthManager
    private(set) var calendarService: GraphCalendarService?
    private(set) var syncService: CalendarSyncService?

    // MARK: - State

    @Published private(set) var isConnected = false
    @Published private(set) var isSyncing = false
    @Published private(set) var lastSyncDate: Date?
    @Published private(set) var cachedEvents: [GraphEvent] = []
    @Published var error: String?

    /// Имя пользователя Outlook
    var userName: String? {
        authManager.userName
    }

    /// Email пользователя Outlook
    var userEmail: String? {
        authManager.userEmail
    }

    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    private init() {
        self.authManager = MSALAuthManager()
        setupObservers()
    }

    private func setupObservers() {
        // Отслеживаем статус авторизации
        authManager.$isSignedIn
            .receive(on: DispatchQueue.main)
            .sink { [weak self] signedIn in
                guard let self = self else { return }

                self.isConnected = signedIn

                if signedIn {
                    // Инициализируем сервисы при подключении
                    self.calendarService = GraphCalendarService(authManager: self.authManager)
                    self.syncService = CalendarSyncService(authManager: self.authManager)
                    self.lastSyncDate = self.syncService?.lastSyncDate

                    // Автоматическая синхронизация при подключении
                    Task {
                        await self.performSync()
                    }
                } else {
                    // Очищаем при отключении
                    self.calendarService = nil
                    self.syncService = nil
                    self.cachedEvents = []
                    self.lastSyncDate = nil
                }
            }
            .store(in: &cancellables)

        // Пробрасываем ошибки из authManager
        authManager.$error
            .receive(on: DispatchQueue.main)
            .assign(to: &$error)
    }

    // MARK: - Connection

    /// Подключить Outlook Calendar
    /// - Parameter viewController: UIViewController для презентации OAuth
    func connect(from viewController: UIViewController) async {
        do {
            _ = try await authManager.signIn(from: viewController)
            debugLog("Connected successfully", module: "OutlookCalendarManager")
        } catch MSALAuthError.userCanceled {
            // Пользователь отменил — не показываем ошибку
            debugLog("User canceled sign in", module: "OutlookCalendarManager", level: .warning)
        } catch {
            self.error = error.localizedDescription
            debugLog("Connection failed: \(error)", module: "OutlookCalendarManager", level: .error)
            debugCaptureError(error, context: "OutlookCalendarManager.connect")
        }
    }

    /// Отключить Outlook Calendar
    /// - Parameter viewController: UIViewController для презентации logout
    func disconnect(from viewController: UIViewController) async {
        do {
            try await authManager.signOut(from: viewController)
            debugLog("Disconnected successfully", module: "OutlookCalendarManager")
        } catch {
            // Fallback на локальный logout
            authManager.signOutLocally()
            debugLog("Fallback to local signout: \(error)", module: "OutlookCalendarManager", level: .warning)
        }
    }

    // MARK: - Sync

    /// Выполнить синхронизацию календаря
    @discardableResult
    func performSync() async -> Bool {
        guard isConnected, let syncService = syncService else {
            debugLog("Cannot sync: not connected", module: "OutlookCalendarManager", level: .warning)
            return false
        }

        guard !isSyncing else {
            debugLog("Sync already in progress", module: "OutlookCalendarManager", level: .warning)
            return false
        }

        isSyncing = true
        error = nil

        defer { isSyncing = false }

        do {
            let result = try await syncService.sync()

            // Обновляем кэш событий
            if result.isFullSync {
                cachedEvents = result.updatedEvents
            } else {
                // Применяем инкрементальные изменения
                applyIncrementalChanges(result)
            }

            lastSyncDate = syncService.lastSyncDate

            debugLog("Sync completed: \(cachedEvents.count) events cached", module: "OutlookCalendarManager")
            return true

        } catch MSALAuthError.interactionRequired {
            self.error = "Требуется повторный вход в Outlook"
            debugLog("Sync failed: interaction required", module: "OutlookCalendarManager", level: .warning)
            return false
        } catch {
            self.error = error.localizedDescription
            debugLog("Sync failed: \(error)", module: "OutlookCalendarManager", level: .error)
            debugCaptureError(error, context: "OutlookCalendarManager.performSync")
            return false
        }
    }

    /// Сбросить и выполнить полную синхронизацию
    func resetAndSync() async {
        syncService?.resetSync()
        await performSync()
    }

    private func applyIncrementalChanges(_ result: SyncResult) {
        // Удаляем удалённые события
        cachedEvents.removeAll { event in
            result.deletedEventIds.contains(event.id)
        }

        // Обновляем/добавляем события
        for updatedEvent in result.updatedEvents {
            if let index = cachedEvents.firstIndex(where: { $0.id == updatedEvent.id }) {
                cachedEvents[index] = updatedEvent
            } else {
                cachedEvents.append(updatedEvent)
            }
        }

        // Сортируем по дате начала
        cachedEvents.sort { ($0.startDate ?? .distantFuture) < ($1.startDate ?? .distantFuture) }
    }

    // MARK: - Events API

    /// Получить события на сегодня
    func getTodayEvents() async throws -> [GraphEvent] {
        guard let service = calendarService else {
            throw OutlookError.notConnected
        }
        return try await service.fetchTodayEvents()
    }

    /// Получить события за период
    func getEvents(from startDate: Date, to endDate: Date) async throws -> [GraphEvent] {
        guard let service = calendarService else {
            throw OutlookError.notConnected
        }
        return try await service.fetchEvents(from: startDate, to: endDate)
    }

    /// Найти событие, перекрывающееся с указанным временем
    /// Полезно для автоматического связывания записи со встречей
    func findOverlappingEvent(at date: Date, duration: TimeInterval) -> GraphEvent? {
        let endDate = date.addingTimeInterval(duration)

        return cachedEvents.first { event in
            guard let eventStart = event.startDate,
                  let eventEnd = event.endDate else { return false }

            // Проверяем перекрытие временных интервалов
            return eventStart <= endDate && eventEnd >= date
        }
    }

    /// Получить ближайшее событие
    func getNextEvent() -> GraphEvent? {
        let now = Date()
        return cachedEvents.first { event in
            guard let eventStart = event.startDate else { return false }
            return eventStart > now
        }
    }

    /// Получить текущее событие (которое идёт сейчас)
    func getCurrentEvent() -> GraphEvent? {
        let now = Date()
        return cachedEvents.first { event in
            guard let eventStart = event.startDate,
                  let eventEnd = event.endDate else { return false }
            return eventStart <= now && eventEnd >= now
        }
    }
}

// MARK: - Errors

enum OutlookError: LocalizedError {
    case notConnected
    case syncFailed(Error)

    var errorDescription: String? {
        switch self {
        case .notConnected:
            return "Outlook Calendar не подключён"
        case .syncFailed(let error):
            return "Ошибка синхронизации: \(error.localizedDescription)"
        }
    }
}

File: Vanta Sppech Android/app/src/main/java/com/vanta/speech/core/calendar/OutlookCalendarManager.kt
package com.vanta.speech.core.calendar

import android.app.Activity
import android.util.Log
import com.vanta.speech.core.calendar.model.GraphEvent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Main coordinator for Outlook Calendar integration
 * Android equivalent of iOS OutlookCalendarManager
 */
@Singleton
class OutlookCalendarManager @Inject constructor(
    val authManager: MSALAuthManager,
    private val calendarService: GraphCalendarService
) {
    companion object {
        private const val TAG = "OutlookCalendarManager"
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // State
    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _isSyncing = MutableStateFlow(false)
    val isSyncing: StateFlow<Boolean> = _isSyncing.asStateFlow()

    private val _lastSyncDate = MutableStateFlow<Date?>(null)
    val lastSyncDate: StateFlow<Date?> = _lastSyncDate.asStateFlow()

    private val _cachedEvents = MutableStateFlow<List<GraphEvent>>(emptyList())
    val cachedEvents: StateFlow<List<GraphEvent>> = _cachedEvents.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    // User info from auth manager
    val userName: StateFlow<String?> = authManager.userName
    val userEmail: StateFlow<String?> = authManager.userEmail

    init {
        // Observe auth state
        scope.launch {
            authManager.isSignedIn.collect { signedIn ->
                _isConnected.value = signedIn

                if (signedIn) {
                    // Auto-sync on connection
                    performSync()
                } else {
                    // Clear on disconnect
                    _cachedEvents.value = emptyList()
                    _lastSyncDate.value = null
                }
            }
        }

        // Forward auth errors
        scope.launch {
            authManager.error.collect { authError ->
                _error.value = authError
            }
        }
    }

    /**
     * Connect to Outlook Calendar
     * @param activity Activity for presenting OAuth UI
     */
    suspend fun connect(activity: Activity) {
        try {
            authManager.signIn(activity)
            Log.d(TAG, "Connected successfully")
        } catch (e: MSALAuthError.UserCanceled) {
            Log.d(TAG, "User canceled sign in")
            // Don't show error for user cancel
        } catch (e: Exception) {
            _error.value = e.message
            Log.e(TAG, "Connection failed: ${e.message}")
        }
    }

    /**
     * Disconnect from Outlook Calendar
     */
    suspend fun disconnect() {
        try {
            authManager.signOut()
            Log.d(TAG, "Disconnected successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Disconnect error: ${e.message}")
            // State is already cleared by signOut
        }
    }

    /**
     * Perform calendar sync
     * @return true if sync was successful
     */
    suspend fun performSync(): Boolean {
        if (!_isConnected.value) {
            Log.w(TAG, "Cannot sync: not connected")
            return false
        }

        if (_isSyncing.value) {
            Log.w(TAG, "Sync already in progress")
            return false
        }

        _isSyncing.value = true
        _error.value = null

        return try {
            // Fetch events for next 7 days
            val calendar = java.util.Calendar.getInstance()
            val startDate = calendar.time
            calendar.add(java.util.Calendar.DAY_OF_MONTH, 7)
            val endDate = calendar.time

            val events = calendarService.fetchEvents(startDate, endDate)

            _cachedEvents.value = events.sortedBy { it.startDate }
            _lastSyncDate.value = Date()

            Log.d(TAG, "Sync completed: ${events.size} events cached")
            true
        } catch (e: MSALAuthError.InteractionRequired) {
            _error.value = "Требуется повторный вход в Outlook"
            Log.w(TAG, "Sync failed: interaction required")
            false
        } catch (e: Exception) {
            _error.value = e.message
            Log.e(TAG, "Sync failed: ${e.message}")
            false
        } finally {
            _isSyncing.value = false
        }
    }

    /**
     * Get today's events
     */
    suspend fun getTodayEvents(): List<GraphEvent> {
        return try {
            calendarService.fetchTodayEvents()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to fetch today's events: ${e.message}")
            emptyList()
        }
    }

    /**
     * Get events for a date range
     */
    suspend fun getEvents(from: Date, to: Date): List<GraphEvent> {
        return try {
            calendarService.fetchEvents(from, to)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to fetch events: ${e.message}")
            emptyList()
        }
    }

    /**
     * Find event overlapping with the specified time
     * Useful for auto-linking recordings to meetings
     */
    fun findOverlappingEvent(at: Date, duration: Long): GraphEvent? {
        val endDate = Date(at.time + duration)

        return _cachedEvents.value.firstOrNull { event ->
            val eventStart = event.startDate ?: return@firstOrNull false
            val eventEnd = event.endDate ?: return@firstOrNull false

            // Check time interval overlap
            eventStart.time <= endDate.time && eventEnd.time >= at.time
        }
    }

    /**
     * Get the next upcoming event
     */
    fun getNextEvent(): GraphEvent? {
        val now = Date()
        return _cachedEvents.value.firstOrNull { event ->
            val eventStart = event.startDate ?: return@firstOrNull false
            eventStart.time > now.time
        }
    }

    /**
     * Get current ongoing event
     */
    fun getCurrentEvent(): GraphEvent? {
        val now = Date()
        return _cachedEvents.value.firstOrNull { event ->
            val eventStart = event.startDate ?: return@firstOrNull false
            val eventEnd = event.endDate ?: return@firstOrNull false
            eventStart.time <= now.time && eventEnd.time >= now.time
        }
    }

    /**
     * Clear any errors
     */
    fun clearError() {
        _error.value = null
        authManager.clearError()
    }
}

/**
 * Outlook integration errors
 */
sealed class OutlookError : Exception() {
    data object NotConnected : OutlookError() {
        private fun readResolve(): Any = NotConnected
        override val message = "Outlook Calendar не подключён"
    }

    data class SyncFailed(override val cause: Throwable) : OutlookError() {
        override val message = "Ошибка синхронизации: ${cause.message}"
    }
}
